Gracefully handle the not_built message from the hashtree.

cmeiklejohn opened the pull request:
Ensure that when aae_source dies, we handle the DOWN message and
terminate the fsssource process instead of ignoring the message.

Do not throw the error, or send the complete message to the sink, when
the process dies; both are not necessary.

cmeiklejohn committed:
Gracefully handle the not_built message from the hashtree.

Ensure that when aae_source dies, we handle the DOWN message and
terminate the fsssource process instead of ignoring the message.

Do not throw the error, or send the complete message to the sink, when
the process dies; both are not necessary.

buddhisthead commented:
Hey Chris, what test do you recommend to reproduce this? Will repl_aae_fullsync show it, if I reduce the max source to 1?

Vagabond commented on src/riak_repl2_fssource.erl:
```diff
@@ -182,6 +182,10 @@ handle_info({'DOWN', Ref, process, _Pid, not_responsible}, State=#state{partitio
         Error -> Error
     end;
+handle_info({'DOWN', Ref, process, _Pid, Reason}, State) ->
+    erlang:demonitor(Ref),
```
You don't need to demonitor a dead process.

Vagabond commented on src/riak_repl2_fssource.erl:
```diff
@@ -182,6 +182,10 @@ handle_info({'DOWN', Ref, process, _Pid, not_responsible}, State=#state{partitio
     end;
+handle_info({'DOWN', Ref, process, _Pid, Reason}, State) ->
+    erlang:demonitor(Ref),
+    lager:info("Received: ~p, stopping process.", [Reason]),
```
If the reason is Error or shutdown, we can probably suppress this.

Vagabond replied:
normal or shutdown, I mean.

Vagabond commented on src/riak_repl_aae_source.erl:
```diff
@@ -170,8 +170,6 @@ prepare_exchange(start_exchange, State=#state{transport=Transport,
         Error ->
             lager:info("AAE source failed get_lock for partition ~p, got ~p",
                        [Partition, Error]),
-            send_complete(State),
```
We should also remove the other throw()s like this, they will all trigger bad_return_type errors. We should also audit the other send_completes to see if they make sense vs just having the socket close.

buddhisthead committed:
Improve handling some more. Remove throw from remote not_built and change return in get_reply.

cmeiklejohn committed:
Fix typos.

buddhisthead committed:
Return error instaed of normal if the aae source fails

buddhisthead committed:
Merge branch 'bugfix/csm/fail-gracefully' of github.com:/basho/riak_repl into bugfix/csm/fail-gracefully

cmeiklejohn committed:
Ensure we start aae_source as linked.

Remove monitors, handle not_responsible as a independent message instead
of a failure state.

cmeiklejohn committed:
Remove unnecessary DOWN clauses.

cmeiklejohn committed:
Use gen_server cast instead of explicit message passing.

cmeiklejohn committed:
Change to handle_cast.

cmeiklejohn committed:
Fix connect in the init.

Vagabond committed:
WIP fullsync partition retry limit

cmeiklejohn committed:
Fix bad API call.

Vagabond committed:
Bound the number of times the fscoordinator will retry a partition

cmeiklejohn committed:
Merge branch 'bugfix/csm/fail-gracefully' of github.com:/basho/riak_repl into bugfix/csm/fail-gracefully

Vagabond committed:
Allow retries exceeded to also trigger fullsync complete

cmeiklejohn committed:
Fail if we can not acquire due to ownership.

cmeiklejohn committed:
Move code which can fail out of the init.

Vagabond committed:
Have the aae source exit if the socket to the sink closes

cmeiklejohn committed:
Unreserve partitions.

Ensure we reserve and unreserve partitions when we detect source
processes exiting.

cmeiklejohn commented:
Closing, incorporated into #500.
